### PR TITLE
Integrate CDC data from MySQL/Postgres/MongoDb data source

### LINKED ISSUE #3233: Support CDC from MySQL
**Is your feature request related to a problem? Please describe.**
I would like to use data-prepper to ingest CDC from MySQL database, including Aurora DB

**Describe the solution you'd like**
Support Kafka Connect with Debezium MySQL connector plugins

**Describe alternatives you've considered (Optional)**
N/A

**Additional context**

```
connect-pipeline:
  source:
    kafka_connect:
      worker_properties:
        group_id: group
        config_storage_topic: pipeline-configs
        offset_storage_topic: pipeline-offsets
        status_storage_topic: pipeline-status
      mysql:
        hostname: localhost
        credentials:
          type: plaintext
          username: username
          password: password
        tables:
          - topic_prefix: prefix1
            table_name: dbname.tableName1
          - topic_prefix: prefix2
            table_name: dbname.tableName2
```

### PR DESCRIPTION
**Pull request by wanghd89**
### Description
 Integrate with Kafka Connect and Debezium MySQL/Postgres/MongoDb plugin for CDC

### Issues Resolved
Resolves #3233 #3234 #3235 
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has a documentation issue. Please link to it in this PR.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mysql-pipeline.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+log-pipeline:
+  source:
+    mysql:
+      hostname: localhost
+      credentials:
+        plaintext:
+          username: debezium
+          password: dbz
+      tables:
+        - topic_prefix: prefix1
```
Comment (sharraj) [inline data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mysql-pipeline.yaml:10]: Do we need this config ? topic_prefix: prefix1. ..Hope we can keep it optional and somehow use say connectorname.table name itself.

File: data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mysql-pipeline.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+          table_name: inventory.customers
```
Comment (sharraj) [inline data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mysql-pipeline.yaml:11]: instead of table_name , can we just say table ?

File: data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-postgres-pipeline.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  source:
+    postgresql:
+      hostname: localhost
+      credentials:
+        plaintext:
+          username: debezium
+          password: dbz
+      tables:
+        - topic_prefix: psql
+          database_name: postgres
```
Comment (sharraj) [inline data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-postgres-pipeline.yaml:11]: instead of database_name, just database.

File: data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-postgres-pipeline.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+          table_name: public.customers
```
Comment (sharraj) [inline data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-postgres-pipeline.yaml:12]: table_name = table.

File: data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mongodb-pipeline.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  source:
+    mongodb:
+      hostname: localhost
+      credentials:
+        plaintext:
+          username: debezium
+          password: dbz
+      collections:
+        - topic_prefix: prefix1
+          collection_name: test.customers
```
Comment (sharraj) [inline data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mongodb-pipeline.yaml:11]: collection_name = collection

File: data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mysql-pipeline.yaml (hunk 1 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/test/resources/sample-mysql-pipeline.yaml:10]: We need this. It is mandatory by Debezium, and it has to be unique across other connectors. Moreover, it is required for OSI as well due to multi-tenant MSK

File: build.gradle [hunk 6]
```diff
@@ -147,13 +147,13 @@ subprojects {
             }
             implementation('org.eclipse.jetty:jetty-http') {
                 version {
-                    require '11.0.15'
+                    require '9.4.51.v20230217'
```
Comment (dlvenable) [inline build.gradle:150]: Can we exclude Jetty from the Debezium project instead?

File: build.gradle [hunk 7]
```diff
@@ -147,13 +147,13 @@ subprojects {
...
                 }
                 because 'CVE from transitive dependencies'
             }
             implementation('org.eclipse.jetty:jetty-server') {
                 version {
-                    require '11.0.15'
+                    require '9.4.51.v20230217'
```
Comment (dlvenable) [inline build.gradle:156]: Can we exclude Jetty from the Debezium project instead?

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/KafkaConnectConfig.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.Properties;
+
+public class KafkaConnectConfig implements KafkaClusterAuthConfig {
+    private static final long CONNECTOR_TIMEOUT_MS = 30000L; // 30 seconds
+    private static final long CONNECT_TIMEOUT_MS = 60000L; // 60 seconds
+
+    @JsonProperty("worker_properties")
+    private WorkerProperties workerProperties = new WorkerProperties();
+
+    @JsonProperty("connect_timeout_ms")
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/KafkaConnectConfig.java:25]: Data Prepper is making duration a first-class primitive in configurations to allow for a few different ways to configure: `30s`, `30000ms`, or even a full ISO-8601 string `PT30S`.

Please rename to `connection_timeout` and use a Java 8 `Duration` type here.

See this PR for an example: #3099

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/configuration/ConnectorConfig.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.opensearch.dataprepper.plugins.kafkaconnect.util.Connector;
+
+import java.util.List;
+import java.util.Properties;
+
+public abstract class ConnectorConfig {
+    @JsonProperty("force_update")
+    public Boolean forceUpdate = false;
+    private String bootstrapServers;
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/configuration/ConnectorConfig.java:17]: Should this be a `List<String>`?

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/WorkerProperties.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    private String configStorageTopic = DEFAULT_CONFIG_STORAGE_TOPIC;
+    @JsonProperty("offset_storage_topic")
+    private String offsetStorageTopic = DEFAULT_OFFSET_STORAGE_TOPIC;
+    @JsonProperty("status_storage_topic")
+    private String statusStorageTopic = DEFAULT_STATUS_STORAGE_TOPIC;
+    @JsonProperty("client_id")
+    private String clientId = DEFAULT_CLIENT_ID;
+    @JsonProperty("offset_storage_partitions")
+    private Integer offsetStoragePartitions = OFFSET_STORAGE_PARTITIONS;
+    @JsonProperty("offset_flush_interval_ms")
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/WorkerProperties.java:51]: See my other comment about using `Duration` instead. This will help yield consistent configurations.

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/SecretManagerHelper.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest;
+
+import java.util.UUID;
+
+public class SecretManagerHelper {
+    private static final String SESSION_PREFIX = "data-prepper-secretmanager-session";
+    public static String getSecretValue(final String stsRoleArn, final String region, final String secretId) {
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/SecretManagerHelper.java:24]: There is some other ongoing work to provide AWS Secrets Manager. See #2780. Let's use that instead.

File: settings.gradle [hunk 12]
```diff
@@ -39,6 +39,8 @@ dependencyResolutionManagement {
             library('bouncycastle-bcpkix', 'org.bouncycastle', 'bcpkix-jdk18on').versionRef('bouncycastle')
             version('guava', '32.0.1-jre')
             library('guava-core', 'com.google.guava', 'guava').versionRef('guava')
+            version('reflections', '0.9.12')
```
Comment (dlvenable) [inline settings.gradle:42]: This downgrade is still a problem. What options do we have to support the latest version?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/KafkaConnectConfig.java (hunk 8 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/KafkaConnectConfig.java:25]: Sure, good to know

Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/KafkaConnectConfig.java:25]: The name should be connect_start_timeout. It means how long should we wait for Kafka Connect in running state.

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/WorkerProperties.java (hunk 10 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/extension/WorkerProperties.java:51]: Sure

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/configuration/ConnectorConfig.java (hunk 9 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/configuration/ConnectorConfig.java:17]: It is String, this config is not used for user input, it is designed for Kafka Connect's worker property, and it is string.
For customer's input, it is List<String>, and I have logic to validate and turn it into string.

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/SecretManagerHelper.java (hunk 11 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/SecretManagerHelper.java:24]: Ok, didn't know that. 
I can do a code clean up to use same SecretManager Utility to get secret after that secret-manager plugin is merged. but for this CR, let's keep it?

Comment (dlvenable) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/SecretManagerHelper.java:24]: I think it would be better to remove it in this PR so that we don't end up losing track of it. For testing purposes you could either:

* Use plain text temporarily
* Keep the changes you have locally.

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSource.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        this.updateConfig(kafkaClusterConfigSupplier);
+        this.kafkaConnect = KafkaConnect.getPipelineInstance(
+                pipelineName,
+                pluginMetrics,
+                kafkaConnectConfig.getConnectStartTimeout(),
+                kafkaConnectConfig.getConnectorStartTimeout());
+    }
+
+    @Override
+    public void start(Buffer<Record<Object>> buffer) {
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSource.java:61]: The `Buffer` isn't used. I understand that this will use the Kafka buffer. But, there should be some tighter connection here. What we don't want is for users to have a Data Prepper buffer that is different from the Kafka Connect topic.

File: data-prepper-plugins/kafka-connect-plugins/build.gradle [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    implementation 'org.eclipse.jetty:jetty-client:9.4.51.v20230217'
+    implementation ('io.confluent:kafka-schema-registry:7.5.0') {
+        exclude group: 'org.glassfish.jersey.containers', module: 'jersey-container-servlet'
+        exclude group: 'org.glassfish.jersey.inject', module: 'jersey-hk2'
+        exclude group: 'org.glassfish.jersey.ext', module: 'jersey-bean-validation'
+    }
+    runtimeOnly 'org.jetbrains.kotlin:kotlin-compiler:1.8.21'
+    // Common Debezium Connector
+    implementation 'io.debezium:debezium-api:2.3.0.Final'
+    implementation 'io.debezium:debezium-core:2.3.0.Final'
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-connect-plugins/build.gradle:31]: It seems that Debezium and Kafka Connect require ZooKeeper. And I see it in the dependency path. How will this work when a user has multiple Data Prepper nodes?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/kafka-connect-plugins/build.gradle [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    implementation 'org.eclipse.jetty:jetty-server:9.4.51.v20230217'
+    implementation 'org.eclipse.jetty:jetty-servlet:9.4.51.v20230217'
+    implementation 'org.eclipse.jetty:jetty-servlets:9.4.51.v20230217'
+    implementation 'org.eclipse.jetty:jetty-client:9.4.51.v20230217'
+    implementation ('io.confluent:kafka-schema-registry:7.5.0') {
+        exclude group: 'org.glassfish.jersey.containers', module: 'jersey-container-servlet'
+        exclude group: 'org.glassfish.jersey.inject', module: 'jersey-hk2'
+        exclude group: 'org.glassfish.jersey.ext', module: 'jersey-bean-validation'
+    }
+    runtimeOnly 'org.jetbrains.kotlin:kotlin-compiler:1.8.21'
```
[the remark this answers is not in the document]
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/build.gradle:28]: The Kafka Connect loads plugins (connectors' code) from java classpath. When running OSDP in my local IDE intellij, it throws 170+ warnning errors like
Caused by: java.lang.ClassNotFoundException: org.jetbrains.kotlin.cli.common.repl.ReplCheckAction
....

but it won't break anything.
You are right, it should not be an issue in OSI or maybe container. Shall we remove it, and leave these errors as known issue if running in local IDE?

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSource.java (hunk 13 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSource.java:61]: We can solve that in documentation/guidance. In OSDP, customers can use their own Kafka as buffer where the topic is filled in by customers. In OSI, we have shared MSK, and the topic is unique. (it is in the design doc)

File: data-prepper-plugins/kafka-connect-plugins/build.gradle (hunk 14 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/build.gradle:31]: https://docs.confluent.io/platform/current/connect/userguide.html#distributed-mode

We make Kafka Connect run in distributed mode, it uses to make multiple nodes work together.
* group.id
* config.storage.topic
* offset.storage.topic 
* status.storage.topic

I cannot find a doc saying Kafka Connect requires Zookeeper to run. Zookeeper works with Kafka itself (not kafka connect).
Kafka Connect needs bootstrap.servers - A list of host/port pairs to use for establishing the initial connection to the Kafka cluster

File: data-prepper-plugins/kafka-connect-plugins/build.gradle (hunk 15 above)
[the remark this answers is not in the document]
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/build.gradle:28]: Just tried the build in docker (follow the steps in develop guidance), the error warnings are there if I remove
```
runtimeOnly 'org.jetbrains.kotlin:kotlin-compiler:1.8.21'
```

**Review by dlvenable — APPROVED** [review]

File: data-prepper-plugins/kafka-connect-plugins/src/test/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSourceTest.java [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+            kafkaConnectSource.start(buffer);
+            verify(kafkaConnect).addConnectors(any());
+            verify(kafkaConnect).start();
+            try {
+                Thread.sleep(10);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+            kafkaConnectSource.stop();
+            verify(kafkaConnect).start();
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/src/test/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSourceTest.java:113]: this should be `verify(kafkaConnect).stop() `?

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSource.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * The abstraction of the kafka connect source.
+ * The kafka connect and connectors are configured and runs async here.
+ */
+@SuppressWarnings("deprecation")
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSource.java:33]: What warnings are suppressed ? Is this related to reflection ?

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+
+        URI advertisedUrl = rest.advertisedUrl();
+        String workerId = advertisedUrl.getHost() + ":" + advertisedUrl.getPort();
+
+        String clientIdBase = ConnectUtils.clientIdBase(config);
+
+        // Create the admin client to be shared by all backing stores.
+        Map<String, Object> adminProps = new HashMap<>(config.originals());
+        ConnectUtils.addMetricsContextProperties(adminProps, config, kafkaClusterId);
+        adminProps.put(CLIENT_ID_CONFIG, clientIdBase + "shared-admin");
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java:186]: the string can be moved to constant.

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+            TimeUnit.MILLISECONDS.sleep(RETRY_INTERVAL_MS);
+        }
+        if (!isRunning) {
+            throw new RuntimeException("Timed out waiting for Kafka Connect running");
+        }
+        LOG.info("Kafka Connect is running");
+    }
+
+    /**
+     * Initialize connectors.
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java:240]: Can you add more documentation on why the below sequence order is important.

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+        ConfigBackingStore configBackingStore = new KafkaConfigBackingStore(
+                internalValueConverter,
+                config,
+                configTransformer,
+                sharedAdmin,
+                clientIdBase);
+
+        // Pass the shared admin to the distributed herder as an additional AutoCloseable object that should be closed when the
+        // herder is stopped. This is easier than having to track and own the lifecycle ourselves.
+        return new DistributedHerder(config, time, worker,
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java:214]: Is there any metrics that we should monitor for this class ?

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+        Set<String> connectorNames = this.connectorMap.keySet();
+        List<String> exceptionMessages = new ArrayList<>();
+        CountDownLatch countDownLatch = new CountDownLatch(connectorNames.size());
+        connectorNames.parallelStream().forEach(connectorName -> {
+            long startTime = clock.millis();
+            boolean isRunning = false;
+            while (clock.millis() - startTime < connectorTimeoutMs) {
+                try {
+                    ConnectorStateInfo info = herder.connectorStatus(connectorName);
+                    if ("RUNNING".equals(info.connector().state())) {
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java:341]: nit: string to constant

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/meter/KafkaConnectMetrics.java [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        String connectorName = name.getKeyProperty("connector");
+        if (connectorName != null) {
+            tags = Tags.concat(tags, "connector", connectorName);
+        }
+
+        String taskName = name.getKeyProperty("task");
+        if (taskName != null) {
+            tags = Tags.concat(tags, "task", taskName);
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/meter/KafkaConnectMetrics.java:139]: nit: more string in this method to constant.

File: data-prepper-plugins/kafka-connect-plugins/README.md [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+- `task-count`: Number of tasks that have run in this worker
+- `connector-count`: The number of connectors that have run in this worker
+- `connector-startup-attempts-total`: Total number of connector startups that this worker has attempted
+- `connector-startup-success-total`: Total number of connector starts that succeeded
+- `connector-startup-failure-total`: Total number of connector starts that failed
+- `task-startup-attempts-total`: Total number of task startups that the worker has attempted
+- `task-startup-success-total`: Total number of task starts that succeeded
+- `task-startup-failure-total`: Total number of task starts that failed
+
+## Connector Metrics
```
Comment (dinujoh) [inline data-prepper-plugins/kafka-connect-plugins/README.md:265]: how is the metrics differentiated for each connector ? is there a prefix/suffix or other dimension ?

**Comment by dinujoh** [comment]
Overall, the changes look good to me. Let's close out the items pointed by @dlvenable and have a plan for that.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaClusterAuthConfig.java [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+import org.opensearch.dataprepper.plugins.kafka.configuration.EncryptionConfig;
+
+public interface KafkaClusterAuthConfig {
+    AwsConfig getAwsConfig();
+
+    AuthConfig getAuthConfig();
+
+    EncryptionConfig getEncryptionConfig();
+
+    String getBootStrapServers();
```
[the remark this answers is not in the document]
Comment (wanghd89) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaClusterAuthConfig.java:19]: This is intended as String. The KafkaClusterAuthConfig is an interface to reuse the KafkaSecurityConfigurer developed by KK. For that it is designed as String

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/extension/KafkaClusterConfig.java [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import jakarta.validation.Valid;
+import org.opensearch.dataprepper.plugins.kafka.configuration.AuthConfig;
+import org.opensearch.dataprepper.plugins.kafka.configuration.AwsConfig;
+import org.opensearch.dataprepper.plugins.kafka.configuration.EncryptionConfig;
+
+import java.util.List;
+
+public class KafkaClusterConfig {
```
[the remark this answers is not in the document]
Comment (wanghd89) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/extension/KafkaClusterConfig.java:16]: The KafkaClusterConfig is designed for users to input configurations. The KafkaClusterAuthConfig is internal common interface to make sure APIs are added to sources where KafkaSecurityConfigurer is needed.

File: data-prepper-plugins/kafka-connect-plugins/src/test/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/ScretManagerHelperTest.java [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+import static org.hamcrest.Matchers.is;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class ScretManagerHelperTest {
```
[the remark this answers is not in the document]
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/test/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/ScretManagerHelperTest.java:32]: Ooops, ty

File: data-prepper-plugins/kafka-connect-plugins/src/test/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSourceTest.java (hunk 16 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/test/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/KafkaConnectSourceTest.java:113]: Yep, ty

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java (hunk 19 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java:240]: Sure

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java (hunk 20 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java:214]: This class is owned by KafkaConnect Runtime. We cannot pass-in metrics object.
Here is the things we can monitor: https://docs.confluent.io/platform/current/connect/monitoring.html
As mentioned in our low-level design review, we can keep add those as needed.

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java (hunk 21 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/util/KafkaConnect.java:341]: Sure

File: data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/meter/KafkaConnectMetrics.java (hunk 22 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/meter/KafkaConnectMetrics.java:139]: ok

File: data-prepper-plugins/kafka-connect-plugins/README.md (hunk 23 above)
Comment (wanghd89) [reply data-prepper-plugins/kafka-connect-plugins/README.md:265]: It is using tags.
Eg.
pipeline_mongodb_source_record_active_count_avg{connector="prefix1.test.customers",serviceName="dataprepper",task="0",} 1341.987951807229

**Review by dlvenable — APPROVED** [review]

**Review by dinujoh — DISMISSED** [review]

**Review by graytaylor0 — APPROVED** [review]

**Review by dinujoh — APPROVED** [review]

**Review by graytaylor0 — APPROVED** [review]
